akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1324,6 +1328,131 @@ private[stream] object Collect {
     }
 }
 
+object AskStage {
+  final class AskStageTargetActorTerminatedException(ref: ActorRef)
+    extends RuntimeException(s"Target Actor of ask() stage has terminated! Was: ${ref}")
+}
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final case class AskStage[In, Out](targetRef: ActorRef, tag: ClassTag[Out], parallelism: Int, timeout: Timeout)
+  extends GraphStage[FlowShape[In, Out]] with AskSupport { stage ⇒
+  require(timeout ne null, "timeout must not be null")
+  import AskStage._
+
+  private val in = Inlet[In]("Ask.in")
+  private val out = Outlet[Out]("Ask.out")
+
+  override def initialAttributes = DefaultAttributes.ask
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler with StageLogging {
+
+      implicit val t: Timeout = timeout // TODO should be also an attribute?
+      implicit def _tag: ClassTag[Out] = tag
+
+      lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider)
+        .getOrElse(Supervision.stoppingDecider)
+
+      private val futureCB = getAsyncCallback[MapAsync.Holder[Out]](holder ⇒
+        holder.elem match {
+          case Success(succ) ⇒
+            pushNextIfPossible()
+
+          case Failure(ex) ⇒
+            holder.supervisionDirectiveFor(decider, ex) match {
+              // fail fast as if supervision says so
+              case Supervision.Stop ⇒ failStage(ex)
+              case _                ⇒ pushNextIfPossible()
+            }
+        })
+
+      private var buffer: BufferImpl[MapAsync.Holder[Out]] = _
+
+      private lazy val self = getStageActor {
+        case (_, Terminated(`targetRef`)) ⇒
+          failStage(new AskStageTargetActorTerminatedException(targetRef))
+      }
+
+      override def preStart(): Unit = {
+        // initialize self, and watch the target
+        self.watch(targetRef)
+
+        buffer = BufferImpl(parallelism, materializer)
+      }
+
+      override def onPull(): Unit = {
+        pushNextIfPossible()
+      }
+
+      override def onPush(): Unit = {
+        try {
+          val elem = grab(in)
+          val future = (targetRef ? elem).map(e ⇒ e.asInstanceOf[Out])(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
```

Comment:
sameThreadExecutionContext for something that can throw is dangerous?

why not use `.mapTo`? you have the ClassTag

Reply:
Seems things go wrong then actually... Rather weird tha then it's racy but it seems to hm...